### PR TITLE
feat: bypass add account confirmations for preinstalled snaps

### PR DESCRIPTION
**Pull request by k-g-j**
## **Description**

This feature bypasses the standard add account confirmations flow for [PREINSTALLED_SNAPS](https://github.com/MetaMask/metamask-extension/blob/feat/preinstalled-snap-confirmations-bypass/app/scripts/snaps/preinstalled-snaps.ts#L4). Optionally, preinstalled snaps may set the `displayConfirmations` flag to `true` when they emit a `"notify: account created"` event, which will then enable them to show the standard add account confirmations.

#### Included
- Skips add account confirmation flow for preinstalled snaps
- Implements options for `displayConfirmation` and `accountNameSuggestion` in snap keyring builder

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24876?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/accounts-planning/issues/407

## **Manual testing steps**

**NO LONGER APPLICABLE** ❌ 
1. You will need to remove `!snap.preinstalled` from the filter in the `getSnapsList` selector [here](https://github.com/MetaMask/metamask-extension/blob/develop/ui/selectors/selectors.js#L2503)
2. Go to three vertical dot menu > Snaps > Account Watcher
3. Input an address or ENS (if on mainnet) into the form and click "Watch account"
4. You should **NOT** see any confirmations screens and see the selected account on the top of the extension change to your newly added address

### **Before**

https://github.com/MetaMask/metamask-extension/assets/91970214/8fc99966-182d-4068-970b-cea641ef4ab5

### **After**

https://github.com/MetaMask/metamask-extension/assets/91970214/8e1460fa-97bd-4a98-a654-53adbcf4032b

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -282,6 +282,7 @@
     "@metamask-institutional/sdk": "^0.1.27",
     "@metamask-institutional/transaction-update": "^0.2.2",
     "@metamask/abi-utils": "^2.0.2",
+    "@metamask/account-watcher": "^3.1.1",
```
Comment (ccharly) [inline package.json:285]: Since it's pre-installed, do we still need it here?

File: ui/selectors/selectors.js [hunk 2]
```diff
@@ -16,81 +16,81 @@ import { NameType } from '@metamask/name-controller';
...
   MOONRIVER_DISPLAY_NAME,
+  NETWORK_TYPES,
+  NetworkStatus,
+  OPTIMISM_DISPLAY_NAME,
+  POLYGON_DISPLAY_NAME,
+  POLYGON_ZKEVM_DISPLAY_NAME,
+  SEPOLIA_DISPLAY_NAME,
+  TEST_CHAINS,
+  TEST_NETWORK_TICKER_MAP,
+  ZK_SYNC_ERA_DISPLAY_NAME,
```
Comment (ccharly) [inline ui/selectors/selectors.js:52]: Since it's sorted, it's hard for me to asses what's changed 🙈 Was it intended? Is there any addition? Is it your IDE's fault? ;)

File: ui/selectors/selectors.js [hunk 3]
```diff
@@ -4,9 +4,9 @@ import { SubjectType } from '@metamask/permission-controller';
 import { ApprovalType } from '@metamask/controller-utils';
```
Comment (ccharly) [inline ui/selectors/selectors.js:4]: This file seems to have been linted or something, is it intended? If not, I would revert it back to its original state. We can lint/do cosmetic changes in another PR probably

File: app/scripts/lib/snap-keyring/snap-keyring.ts [hunk 4]
```diff
@@ -1,18 +1,20 @@
+// noinspection t
+
```
Comment (ccharly) [inline app/scripts/lib/snap-keyring/snap-keyring.ts:2]: Probably better if we remove this. Is it for your IDE?

File: app/scripts/lib/snap-keyring/snap-keyring.ts [hunk 5]
```diff
@@ -141,19 +144,29 @@ export const snapKeyringBuilder = (
         const learnMoreLink =
           'https://support.metamask.io/hc/en-us/articles/360015289452-How-to-add-accounts-in-your-wallet';
 
+        // If snap is preinstalled and does not request confirmation, skip the confirmation dialog
+        const skipConfirmation =
+          isSnapPreinstalled(snapId) && !displayConfirmation;
         // Since we use this in the finally, better to give it a default value if the controller call fails
-        let confirmationResult = false;
+        let confirmationResult = skipConfirmation;
```
Comment (ccharly) [inline app/scripts/lib/snap-keyring/snap-keyring.ts:151]: The comment does not really "match" anymore, maybe something like:

```ts
// If confirmation dialog are skipped, we consider the account creation to be confirmed  
```

WDYT?

File: app/scripts/lib/snap-keyring/snap-keyring.ts [hunk 6]
```diff
@@ -141,19 +144,29 @@ export const snapKeyringBuilder = (
...
+        let addAccountApprovalId = '';
```
Comment (ccharly) [inline app/scripts/lib/snap-keyring/snap-keyring.ts:152]: I feel like `confirmationApprovalId` would be a bit better and "aligned" with the `confirmationResult`, WDYT?

File: package.json (hunk 1 above)
Comment (k-g-j) [reply package.json:285]: Yes cause we need to grab the `preinstalled-snap.json` from the npm package [here](https://github.com/MetaMask/metamask-extension/blob/feat/preinstalled-snap-confirmations-bypass/app/scripts/snaps/preinstalled-snaps.ts#L3) :)

File: ui/selectors/selectors.js (hunk 2 above)
Comment (k-g-j) [reply ui/selectors/selectors.js]: It was my IDE doing automatic linting! 😮‍💨 I fixed it here [4b2ae4b](https://github.com/MetaMask/metamask-extension/pull/24876/commits/4b2ae4b597e91de9850b4f81ced25eed86ffab25)

File: ui/selectors/selectors.js (hunk 3 above)
Comment (k-g-j) [reply ui/selectors/selectors.js]: It was my IDE doing automatic linting! 😮‍💨 I fixed it here [4b2ae4b](https://github.com/MetaMask/metamask-extension/pull/24876/commits/4b2ae4b597e91de9850b4f81ced25eed86ffab25)

File: app/scripts/lib/snap-keyring/snap-keyring.ts (hunk 5 above)
Comment (k-g-j) [reply app/scripts/lib/snap-keyring/snap-keyring.ts]: Good catch! [e448aa4](https://github.com/MetaMask/metamask-extension/pull/24876/commits/e448aa4d896df0b93b46f5fa140b752078f581e7)

File: app/scripts/lib/snap-keyring/snap-keyring.ts (hunk 6 above)
Comment (k-g-j) [reply app/scripts/lib/snap-keyring/snap-keyring.ts]: Nice call! [44bb44a](https://github.com/MetaMask/metamask-extension/pull/24876/commits/44bb44a2ce8a7f924671ef81cd2efca85d87ee98)

**Comment by k-g-j** [comment]
@SocketSecurity ignore npm/ethers@6.12.1

File: app/scripts/lib/snap-keyring/snap-keyring.ts [hunk 7]
```diff
@@ -41,6 +41,7 @@ export const getAccountsBySnapId = async (
  * @param trackEvent - A function to track MetaMetrics events.
  * @param getSnapName - A function to get a snap's localized
  * (or non-localized if there are no localization files) name from its manifest.
+ * @param isSnapPreinstalled
```
Comment (ccharly) [inline app/scripts/lib/snap-keyring/snap-keyring.ts:44]:
```suggestion
 * @param isSnapPreinstalled - A function to check if a Snap is pre-installed.
```

File: ui/selectors/selectors.js [hunk 8]
```diff
@@ -2470,10 +2470,7 @@ export function getIsDesktopEnabled(state) {
 export function getSnapsList(state) {
   const snaps = getSnaps(state);
   return Object.entries(snaps)
-    .filter(
-      ([_key, snap]) =>
-        !snap.preinstalled && snap.status !== SnapStatus.Installing,
-    )
+    .filter(([_key, snap]) => snap.status !== SnapStatus.Installing)
```
Comment (ccharly) [inline ui/selectors/selectors.js:2473]: Is this really related to this PR? I guess it's required to trigger the watch-only account creation from the Snaps menu, but do we want to have this as a new behavior?

File: app/scripts/snaps/preinstalled-snaps.ts [hunk 9]
```diff
@@ -1,8 +1,10 @@
 import type { PreinstalledSnap } from '@metamask/snaps-controllers';
 import MessageSigningSnap from '@metamask/message-signing-snap/dist/preinstalled-snap.json';
+import AccountWatcherSnap from '@metamask/account-watcher/dist/preinstalled-snap.json';
```
Comment (ccharly) [inline app/scripts/snaps/preinstalled-snaps.ts:3]: Same questions than for `getSnapsList`, if we plan to have this new Snap as a pre-installed one from now on, I wonder if this should be in a separate PR.

(This is really an open-question, and maybe this was the plan to have it all bundle together 😄)

File: ui/selectors/selectors.js (hunk 8 above)
Comment (k-g-j) [reply ui/selectors/selectors.js:2501]: It is not necessary for the PR. I can remove it but then I will have to add to disable that selector in the manual testing steps.

File: app/scripts/snaps/preinstalled-snaps.ts (hunk 9 above)
Comment (k-g-j) [reply app/scripts/snaps/preinstalled-snaps.ts:3]: I would like to add the account watcher here because we can also use it to test the `accountNameSuggestion` once we tackle that, but I can remove the changes to the selector and add to disable it in the manual testing steps

**Review by ccharly — DISMISSED** [review]

File: app/scripts/lib/snap-keyring/snap-keyring.ts [hunk 10]
```diff
@@ -141,19 +142,29 @@ export const snapKeyringBuilder = (
         const learnMoreLink =
           'https://support.metamask.io/hc/en-us/articles/360015289452-How-to-add-accounts-in-your-wallet';
 
-        // Since we use this in the finally, better to give it a default value if the controller call fails
-        let confirmationResult = false;
+        // If snap is preinstalled and does not request confirmation, skip the confirmation dialog
+        const skipConfirmation =
+          isSnapPreinstalled(snapId) && !displayConfirmation;
+        // If confirmation dialog are skipped, we consider the account creation to be confirmed
```
Comment (gantunesr) [inline app/scripts/lib/snap-keyring/snap-keyring.ts:148]:
```suggestion
        // If confirmation dialog is skipped, we consider the account creation to be confirmed
```

**Review by gantunesr — DISMISSED** [review]
LGTM, just a nit that can be solved in another PR to not block this one

**Review by gantunesr — APPROVED** [review]

**Review by ccharly — APPROVED** [review]
